### PR TITLE
Lorentz correction

### LINKED ISSUE #18: Implement Lorentz correction
See requirement on Confluence: https://confluence.esss.lu.se/display/DAM/DREAM+-+Lorentz+correction

### PR DESCRIPTION
**Pull request by jl-wynen**
Fixes #18

### CODE REVIEW

File: src/ess/diffraction/powder/correction.py [hunk 1]
```diff
@@ -44,3 +49,68 @@ def merge_calibration(*, into: sc.DataArray, calibration: sc.Dataset) -> sc.Data
...
+    :
+        Powder Lorentz factor.
+    """
+
+    def f(dspacing: sc.Variable, two_theta: sc.Variable) -> sc.Variable:
+        return dspacing**4 * sc.sin(two_theta / 2)
+
+    aux = data.transform_coords(
+        'lorentz_factor', {'lorentz_factor': f}, rename_dims=False, quiet=True
+    )
```
Comment (SimonHeybrock) [inline src/ess/diffraction/powder/correction.py:88]: This entire function is odd, it seems `transform_coords` is only used to extract either a binned or regular coord, but then we have to have the special handling below? Shouldn't we either to a single `transform_coords` that also computes `two_theta`, or not have this helper at all?

File: src/ess/diffraction/powder/correction.py [hunk 2]
```diff
@@ -44,3 +49,68 @@ def merge_calibration(*, into: sc.DataArray, calibration: sc.Dataset) -> sc.Data
...
+        Input data with coordinates ``two_theta`` and ``dspacing``.
+
+    Returns
+    -------
+    :
+        Powder Lorentz factor.
+    """
+
+    def f(dspacing: sc.Variable, two_theta: sc.Variable) -> sc.Variable:
+        return dspacing**4 * sc.sin(two_theta / 2)
```
[the remark this answers is not in the document]
Comment (jl-wynen) [reply src/ess/diffraction/powder/correction.py:84]: In scippneutron, we typically have a 'main' coord (tof, wavelength, ...) that determines the dtype. Here, I don't see a hierarchy. So I would expect the normal type promotion behaviour.

File: src/ess/diffraction/powder/correction.py [hunk 3]
```diff
@@ -44,3 +49,68 @@ def merge_calibration(*, into: sc.DataArray, calibration: sc.Dataset) -> sc.Data
...
+
+    aux = data.transform_coords(
+        'lorentz_factor', {'lorentz_factor': f}, rename_dims=False, quiet=True
+    )
+    if aux.bins is not None:
+        return aux.bins.coords['lorentz_factor']
```
[the remark this answers is not in the document]
Comment (jl-wynen) [reply src/ess/diffraction/powder/correction.py:90]: Given https://github.com/scipp/scipp/issues/3392, the bin coords cannot be stripped. I have to create a new data array from scratch.

[the remark this answers is not in the document]
Comment (jl-wynen) [reply src/ess/diffraction/powder/correction.py:90]: But yes, it is a mistake to return only the variable.

File: src/ess/diffraction/powder/correction.py [hunk 4]
```diff
@@ -44,3 +49,68 @@ def merge_calibration(*, into: sc.DataArray, calibration: sc.Dataset) -> sc.Data
...
+    See :func:`ess.diffraction.power.correction.lorentz_factor`.
+    """
+    # The pipeline does not compute two_theta because it uses a conversion
+    # with calibration to compute dspacing.
+    data = data.transform_coords(
+        'two_theta', graph=beamline(scatter=True), quiet=True, rename_dims=False
+    )
+
+    factor = lorentz_factor(data)
+    return LorentzCorrectedData[RunType](factor * data)
```
Comment (SimonHeybrock) [inline src/ess/diffraction/powder/correction.py:112]: Maybe something like this:
```python
# shallow copy da and bins content
dspacing = da.bins.coords['dspacing'] is da.bins is not None
two_theta = da.bins.coords['two_theta'] if 'two_theta' in da.bins.coords else da.coords['two_theta']
# dspacing**4 * sc.sin(two_theta / 2)
theta = 0.5 * two_theta
sin_theta = sc.sin(theta, out=theta)
da.bins.data = dspacing ** 4
da.bins.data *= orig.bins.data
da.bins.data *= sin_theta
```

File: tests/diffraction/powder/corrections_test.py [hunk 5]
```diff
@@ -109,3 +111,157 @@ def test_merge_calibration_raises_if_mask_exists(calibration):
...
+        sc.testing.assert_identical(coord, original.coords[key])
+
+
+def test_lorentz_factor_dense_2d_coord():
+    da = sc.DataArray(
+        sc.full(value=0.7, sizes={'detector_number': 3, 'dspacing': 4}),
+        coords={
+            'dspacing': sc.array(
+                dims=['dspacing'], values=[0.1, 0.4, 0.7, 1.1], unit='angstrom'
+            ).broadcast(sizes={'detector_number': 3, 'dspacing': 4}),
```
Comment (YooSunYoung) [inline tests/diffraction/powder/corrections_test.py:177]: Can we move this ``broadcast`` to line#195 instead?

Comment (jl-wynen) [reply tests/diffraction/powder/corrections_test.py:177]: No. This test is specifically about testing with a 2d input coordinate.

File: src/ess/diffraction/powder/correction.py [hunk 6]
```diff
@@ -44,3 +46,67 @@ def merge_calibration(*, into: sc.DataArray, calibration: sc.Dataset) -> sc.Data
         )
     out.masks['calibration'] = calibration['mask'].data
     return out
+
+
+def lorentz_correction(da: sc.DataArray) -> sc.DataArray:
```
Comment (SimonHeybrock) [inline src/ess/diffraction/powder/correction.py:51]:
```suggestion
def apply_lorentz_correction(da: sc.DataArray) -> sc.DataArray:
```
or something like that, given that we changed what this function does?

File: src/ess/diffraction/powder/correction.py [hunk 7]
```diff
@@ -44,3 +46,67 @@ def merge_calibration(*, into: sc.DataArray, calibration: sc.Dataset) -> sc.Data
...
+    Parameters
+    ----------
+    da:
+        Input data with coordinates ``two_theta`` and ``dspacing``.
+
+    Returns
+    -------
+    :
+        ``da`` multiplied by :math:`L`.
+        has the same dtype as ``da``.
```
Comment (SimonHeybrock) [inline src/ess/diffraction/powder/correction.py:78]:
```suggestion
        Has the same dtype as ``da``.
```

File: src/ess/diffraction/powder/correction.py [hunk 8]
```diff
@@ -44,3 +46,67 @@ def merge_calibration(*, into: sc.DataArray, calibration: sc.Dataset) -> sc.Data
...
+    dspacing = _event_or_bin_coord(da, 'dspacing')
+    two_theta = _event_or_bin_coord(da, 'two_theta')
+    theta = 0.5 * two_theta
+
+    d4 = dspacing.broadcast(sizes=out.sizes) ** 4
+    if out.bins is None:
+        out.data = d4.to(dtype=out.dtype, copy=False)
+        out_data = out.data
+    else:
+        out.bins.data = d4.to(dtype=out.bins.dtype)
```
Comment (SimonHeybrock) [inline src/ess/diffraction/powder/correction.py:92]:
```suggestion
        out.bins.data = d4.to(dtype=out.bins.dtype, copy=False)
```
or is that not possible?

File: src/ess/diffraction/powder/correction.py [hunk 9]
```diff
@@ -44,3 +46,67 @@ def merge_calibration(*, into: sc.DataArray, calibration: sc.Dataset) -> sc.Data
...
+    out = da.copy(deep=False)
+    if da.bins is not None:
+        out.data = sc.bins(**da.bins.constituents)
+    return out
+
+
+def _event_or_bin_coord(da: sc.DataArray, name: str) -> sc.Variable:
+    try:
+        return da.bins.coords[name]
+    except (AttributeError, KeyError):
```
Comment (SimonHeybrock) [inline src/ess/diffraction/powder/correction.py:110]: Do we have a test that ensures that this always favors the event-coord? We have to make sure this is not broken in the future.

Comment (jl-wynen) [reply src/ess/diffraction/powder/correction.py:110]: Added

**Review by SimonHeybrock — APPROVED** [review]

**Pull request merged**